### PR TITLE
[Tenable Security Center]: Connector Initial Creation

### PR DESCRIPTION
**Pull request by flavienSindou**
### Proposed changes

* Implement Tenable Security Center External Import Connector

### Related issues

* https://github.com/OpenCTI-Platform/connectors/issues/2847

### Checklist

- [x] I consider the submitted work as finished
- [x] I tested the code for its functionality using different use cases
- [x] I added/update the relevant documentation (either on github or on notion)
- [ ] Where necessary I refactored code to improve the overall quality

### Further comments
 **Source**

**Tenable Security Center** is an on-premises solution for vulnerability management developed by Tenable. It uses `Plugins` to scan `Assets` and identify `Findings`.

 **Target**

The developed connector retrieves data from Tenable APIs, processes it, and generates compromised `Systems` based on the detected `Vulnerabilities` and related `Observables` (IP addresses, MAC addresses, domain names, operating systems, and software).

- Simplified data model
    
    ```mermaid
    graph LR
        subgraph TenableSecurityCenter
            direction TB
            TenableHostAsset[Asset]
            TenableVulnerability[Finding]
        end
    
        subgraph OpenCTI
            direction LR
            subgraph DomainObjects
                direction TB
                OpenCTISystem[System]
                OpenCTIVulnerability[Vulnerabilities]
            end
    
            subgraph Observables
                direction TB
                OpenCTIMACAddress[MACAddress]
                OpenCTIPv4Address[IPAddress v4]
                OpenCTIPv6Address[IPAddress v6]
                OpenCTIHostname[Hostname]
                OpenCTIDomainName[DomainName]
                OpenCTIOperatingSystem[OperatingSystems]
                OpenCTISoftware[Software_s]
            end
            %% OpenCTIMACAddress ~~~ OpenCTIPv4Address
        end
    
        %% Asset generates System and Observables
        TenableHostAsset ==> OpenCTISystem & OpenCTIMACAddress & OpenCTIPv4Address & OpenCTIPv6Address & OpenCTIHostname & OpenCTIDomainName
        TenableHostAsset==>|looping over installed OS| OpenCTIOperatingSystem
    
        %% Plugin generates Software and Vulnerability
        TenableVulnerability ==>|looping over targeted CPE URIs| OpenCTISoftware
        TenableVulnerability ==>|looping over detected CVEs| OpenCTIVulnerability
    
        %% Relationships between System and Observables
        OpenCTISystem -.-> |"Related to"| OpenCTIMACAddress & OpenCTIPv4Address & OpenCTIPv6Address & OpenCTIHostname & OpenCTIDomainName & OpenCTIOperatingSystem
        %% Relationships between observables
        OpenCTIDomainName -.->|"Resolves to"| OpenCTIPv4Address & OpenCTIPv6Address -.->|"Resolves to"| OpenCTIMACAddress
        
        %% Relationships between Vulnerabilities and Software
        OpenCTISoftware -.-> |"Has"| OpenCTIVulnerability
        OpenCTISystem -.-> |"Has"| OpenCTIVulnerability
    ```
    

**Known Issues**
- The lack of filtering options in the API
- Variations in API responses necessitated a resilient port/adapter architecture to handle changes.
- the possibly huge amount of data to ingest.

**Technical Feature**
- Multi-stage Docker builds.
- A dedicated Python package to encapsulate core logic.
- Configuration through environment variables to avoid embedding sensitive data and access to disk memory.
- Multi-processing via threads

### CODE REVIEW

File: external-import/tenable-security-center/tenable_security_center/domain/entities.py
Comment (Powlinett) [inline external-import/tenable-security-center/tenable_security_center/domain/entities.py]: `_stix2_representation` is not used for ExternalReference instances so it can be removed

File: external-import/tenable-security-center/tenable_security_center/adapters/tsc_api/v5_13_common.py
Comment (Powlinett) [inline external-import/tenable-security-center/tenable_security_center/adapters/tsc_api/v5_13_common.py]: after second thought, it might be better to move these filters into README

**Review by Powlinett — COMMENTED** [review]
✅ pair-reviewed + second review on my own
✅tests pass
✅docker image is building correctly
✅no unexpected errors during ingestion

Waiting for you last commits to approve and merge it 🙏

File: external-import/tenable-security-center/tenable_security_center/adapters/tsc_api/v5_13_common.py
Comment (flavienSindou) [reply external-import/tenable-security-center/tenable_security_center/adapters/tsc_api/v5_13_common.py]: Thanks for you remark, As it is likely to evolve I removed it

**Review by Powlinett — APPROVED** [review]

File: external-import/tenable-security-center/app.py
Comment (helene-nguyen) [inline external-import/tenable-security-center/app.py]: Thanks for this huge work!
Could you add here `cleanup_inconsistent_bundle` set to `True`? :)
Since >= 6.3.3, when using send_stix2_bundle, for having a consistent bundle when splitting the data, it will prevent from MISSING_ERROR_REFERENCE when creating a relationship for example

Here are some example of cases

image


```suggestion
        bundles_sent = self.helper.send_stix2_bundle(
            bundle=bundle_json,
            work_id=self.work_id,
            cleanup_inconsistent_bundle=True
        )
```

**Review by helene-nguyen — CHANGES_REQUESTED** [review]

File: external-import/tenable-security-center/app.py
Comment (flavienSindou) [reply external-import/tenable-security-center/app.py]: Thank you for your feedback.

Given the potential volume of duplicated data (particularly with CVE and software (CPE) ingestions), the bundle was intentionally made inconsistent using a caching mechanism to reduce queue saturation. For example, the author, a previously sent CVE, or a previously sent targeted software will not be resent during the same "work" process.

Testing with sample data showed this deduplication approach reduced the volume of data to ingest by roughly a factor of three.

Does this methodology and optimization meet your expectations, or do you believe that maintaining a fully consistent bundle should take priority ?

Comment (helene-nguyen) [reply external-import/tenable-security-center/app.py]: Thanks for your detailed explanation.

The caching mechanism that you are talking about is not in the client-python helper, right ?

As we want to have consistent bundle when the worker consume the RabbitMQ queue, I'm interested to talk about it in details.

Comment (flavienSindou) [reply external-import/tenable-security-center/app.py]: @richard-julien supports the drop of deduplication solution

**Review by helene-nguyen — APPROVED** [review]

**Review by helene-nguyen — COMMENTED** [review]
As you add adaption for reading config.yml, could you add the sample too ?

**Pull request merged**
